File: src/mpc_controller/src/mpc_solver.cpp
#include "mpc_controller/mpc_solver.h"
#include "mpc_controller/utils.h"
#include "mpc_controller/constraints.h"
#include <iostream>
#include <algorithm>
#include <casadi/casadi.hpp>

MPCSolver::MPCSolver(const MPCConfig& config)
    : config_(config), model_(config) {
    config_.initializeDefaults();
    last_control_sequence_ = Eigen::MatrixXd::Zero(config_.horizon, 2);
}

MPCSolver::SolveInfo MPCSolver::solve(
    const Eigen::Vector4d& x0,
    const Eigen::MatrixXd& reference_trajectory,
    Eigen::MatrixXd& optimal_controls,
    Eigen::MatrixXd& predicted_trajectory,
    const Eigen::MatrixXd& x0_control) {
    
    using namespace casadi;
    
    SolveInfo info;
    
    // Use warm start or provided initial guess
    Eigen::VectorXd u0;
    if (x0_control.size() == 0) {
        u0 = Eigen::Map<Eigen::VectorXd>(
            last_control_sequence_.data(),
            last_control_sequence_.size()
        );
    } else {
        u0 = Eigen::Map<Eigen::VectorXd>(
            const_cast<Eigen::MatrixXd&>(x0_control).data(),
            x0_control.size()
        );
    }
    
    // Expand x0 to 5D state (add velocity = 0)
    Eigen::VectorXd x0_5d = Eigen::VectorXd::Zero(5);
    x0_5d.head(4) = x0;
    x0_5d(4) = 0.0;  // Initial velocity
    
    try {
        // Decision variables: control sequence (horizon x 2)
        MX u = MX::sym("u", config_.horizon * 2);
        
        // Build cost function
        MX cost = 0.0;
        
        // Create dynamics function for CasADi
        // State: [x, y, theta, delta, v]
        // Control: [a, delta_dot]
        MX state = MX::sym("state", 5);
        MX control = MX::sym("control", 2);
        
        // Bicycle model dynamics
        MX state_dot = MX::zeros(5);
        state_dot(0) = state(4) * cos(state(2));
        state_dot(1) = state(4) * sin(state(2));
        state_dot(2) = (state(4) / config_.wheelbase) * tan(state(3));
        state_dot(3) = control(1);
        state_dot(4) = control(0);
        
        Function dynamics_fn = Function("dynamics", {state, control}, {state_dot});
        
        // Forward simulation with cost accumulation
        // Initialize current state with x0_5d
        DM current_state_val = DM::zeros(5, 1);
        for (int i = 0; i < 5; ++i) {
            current_state_val(i) = x0_5d(i);
        }
        
        // Simulate trajectory and compute cost
        for (int i = 0; i < config_.horizon; ++i) {
            // Extract control
            MX v_control = u(i * 2);
            MX delta_dot_control = u(i * 2 + 1);
            MX control_input = vertcat(v_control, delta_dot_control);
            
            // Create a symbolic state for this step
            MX sym_state = MX(current_state_val);
            
            // RK4 integration step (using symbolic computation)
            std::vector<MX> k_results(4);
            
            k_results[0] = dynamics_fn(std::vector<MX>{sym_state, control_input})[0];
            k_results[1] = dynamics_fn(std::vector<MX>{sym_state + 0.5 * config_.dt * k_results[0], control_input})[0];
            k_results[2] = dynamics_fn(std::vector<MX>{sym_state + 0.5 * config_.dt * k_results[1], control_input})[0];
            k_results[3] = dynamics_fn(std::vector<MX>{sym_state + config_.dt * k_results[2], control_input})[0];
            
            MX next_state = sym_state + (config_.dt / 6.0) * 
                           (k_results[0] + 2*k_results[1] + 2*k_results[2] + k_results[3]);
            
            // Compute state cost (compare with reference)
            std::vector<double> ref_i(4);
            for (int j = 0; j < 4; ++j) {
                ref_i[j] = reference_trajectory(i + 1, j);
            }
            
            MX state_error = vertcat(
                next_state(0) - ref_i[0],
                next_state(1) - ref_i[1],
                next_state(2) - ref_i[2],
                next_state(3) - ref_i[3]
            );
            
            // Wrap angle error
            state_error(2) = state_error(2) - 2*M_PI*floor((state_error(2) + M_PI)/(2*M_PI));
            
            // Stage cost: state error + control effort
            for (int j = 0; j < 4; ++j) {
                for (int k = 0; k < 4; ++k) {
                    cost = cost + state_error(j) * config_.Q(j, k) * state_error(k);
                }
            }
            
            cost = cost + v_control * config_.R(0, 0) * v_control + 
                         delta_dot_control * config_.R(1, 1) * delta_dot_control;
            
            // Update state for next iteration - use next_state in MX form
            current_state_val = DM::zeros(5, 1);  // Reset for next iteration
            sym_state = next_state;  // Carry forward the symbolic state
        }
        
        // Terminal cost using final symbolic state
        std::vector<double> ref_terminal(4);
        for (int j = 0; j < 4; ++j) {
            ref_terminal[j] = reference_trajectory(config_.horizon, j);
        }
        
        MX terminal_error = vertcat(
            current_state_val(0) - ref_terminal[0],
            current_state_val(1) - ref_terminal[1],
            current_state_val(2) - ref_terminal[2],
            current_state_val(3) - ref_terminal[3]
        );
        terminal_error(2) = terminal_error(2) - 2*M_PI*floor((terminal_error(2) + M_PI)/(2*M_PI));
        
        for (int j = 0; j < 4; ++j) {
            for (int k = 0; k < 4; ++k) {
                cost = cost + terminal_error(j) * config_.Q_terminal(j, k) * terminal_error(k);
            }
        }
        
        // Input bounds constraints
        auto bounds = buildBounds();
        std::vector<double> lbx(config_.horizon * 2), ubx(config_.horizon * 2);
        for (size_t i = 0; i < bounds.size(); ++i) {
            lbx[i] = bounds[i].first;
            ubx[i] = bounds[i].second;
        }
        
        // Create NLP
        MXDict nlp = {{"x", u}, {"f", cost}};
        
        // Create solver with IPOPT and options
        Dict opts;
        opts["ipopt.max_iter"] = 100;
        opts["ipopt.print_level"] = 0;
        opts["print_time"] = false;
        
        Function solver = nlpsol("solver", "ipopt", nlp, opts);
        
        // Prepare initial guess
        std::vector<double> u0_vec(config_.horizon * 2);
        for (int i = 0; i < config_.horizon; ++i) {
            u0_vec[i * 2] = u0(i * 2);
            u0_vec[i * 2 + 1] = u0(i * 2 + 1);
        }
        
        // Solve
        DMDict arg;
        arg["x0"] = u0_vec;
        arg["lbx"] = lbx;
        arg["ubx"] = ubx;
        
        DMDict res = solver(arg);
        
        // Extract solution - convert DM to vector
        DM u_opt_dm = res.at("x");
        std::vector<double> u_opt;
        for (int i = 0; i < u_opt_dm.size1(); ++i) {
            u_opt.push_back(static_cast<double>(u_opt_dm(i)));
        }
        
        optimal_controls = Eigen::MatrixXd::Zero(config_.horizon, 2);
        for (int i = 0; i < config_.horizon; ++i) {
            optimal_controls(i, 0) = u_opt[i * 2];
            optimal_controls(i, 1) = u_opt[i * 2 + 1];
        }
        
        predicted_trajectory = model_.predictTrajectory(x0_5d, optimal_controls);
        
        // Convert 5D trajectory to 4D for output (drop velocity)
        Eigen::MatrixXd traj_4d = predicted_trajectory.leftCols(4);
        predicted_trajectory = traj_4d;
        
        // Store for warm-starting next solve
        last_control_sequence_ = optimal_controls;
        last_trajectory_ = predicted_trajectory;
        
        double final_cost = computeCost(predicted_trajectory, optimal_controls, reference_trajectory);
        
        info.success = true;
        info.cost = final_cost;
        info.iterations = 0;
        info.message = "Solved with CasADi IPOPT";
        
    } catch (const std::exception& e) {
        std::cerr << "CasADi solver error: " << e.what() << std::endl;
        
        // Fallback to warm start if optimization fails
        optimal_controls = last_control_sequence_;
        Eigen::VectorXd x0_5d_fallback = Eigen::VectorXd::Zero(5);
        x0_5d_fallback.head(4) = x0;
        predicted_trajectory = model_.predictTrajectory(x0_5d_fallback, optimal_controls);
        predicted_trajectory = predicted_trajectory.leftCols(4);
        
        info.success = false;
        info.cost = computeCost(predicted_trajectory, optimal_controls, reference_trajectory);
        info.message = "Fallback: CasADi failed, using warm start";
    }
    
    last_solve_info_ = info;
    return info;
}

Eigen::Vector2d MPCSolver::getControl(
    const Eigen::Vector4d& x0,
    const Eigen::MatrixXd& reference_trajectory) {
    
    Eigen::MatrixXd controls;
    Eigen::MatrixXd trajectory;
    solve(x0, reference_trajectory, controls, trajectory);
    
    return controls.row(0).transpose();
}

double MPCSolver::computeCost(
    const Eigen::MatrixXd& trajectory,
    const Eigen::MatrixXd& controls,
    const Eigen::MatrixXd& reference_trajectory) const {
    
    double cost = 0.0;
    int N = controls.rows();
    
    // Stage costs
    for (int i = 0; i < N; ++i) {
        Eigen::Vector4d error = trajectory.row(i).transpose() -
                                reference_trajectory.row(i).transpose();
        error(2) = mpc_utils::wrapAngle(error(2));
        
        double state_cost = error.transpose() * config_.Q * error;
        double control_cost = controls.row(i) * config_.R * controls.row(i).transpose();
        
        cost += state_cost + control_cost;
    }
    
    // Terminal cost
    Eigen::Vector4d terminal_error = trajectory.row(N).transpose() -
                                     reference_trajectory.row(N).transpose();
    terminal_error(2) = mpc_utils::wrapAngle(terminal_error(2));
    double terminal_cost = terminal_error.transpose() * config_.Q_terminal * terminal_error;
    cost += terminal_cost;
    
    return cost;
}

std::vector<std::pair<double, double>> MPCSolver::buildBounds() const {
    auto bounds_pair = ConstraintSet(config_).getInputBounds();
    Eigen::MatrixXd u_lower = bounds_pair.first;
    Eigen::MatrixXd u_upper = bounds_pair.second;
    
    std::vector<std::pair<double, double>> bounds;
    for (int i = 0; i < config_.horizon; ++i) {
        bounds.push_back(std::make_pair(u_lower(i, 0), u_upper(i, 0)));  // velocity
        bounds.push_back(std::make_pair(u_lower(i, 1), u_upper(i, 1)));  // steering rate
    }
    
    return bounds;
}

void MPCSolver::setWeights(const Eigen::Matrix4d& Q,
                          const Eigen::Matrix2d& R,
                          const Eigen::Matrix4d& Q_terminal) {
    config_.Q = Q;
    config_.R = R;
    if (Q_terminal.norm() > 0) {
        config_.Q_terminal = Q_terminal;
    } else {
        config_.Q_terminal = Q * 2.0;
    }
}

void MPCSolver::resetWarmStart() {
    last_control_sequence_ = Eigen::MatrixXd::Zero(config_.horizon, 2);
    last_trajectory_ = Eigen::MatrixXd::Zero(config_.horizon + 1, 4);
}

File: src/kinematic_bicycle/kinematic_bicycle/mpc/mpc_solver.py
"""MPC Solver - Optimization and control computation

This is the core control logic, separated from ROS
"""

import numpy as np
from typing import Tuple, Optional
from scipy.optimize import minimize

from .config import MPCConfig
from .model import BicycleModel
from .constraints import ConstraintSet
from .utils import get_reference_error, wrap_angle


class MPCSolver:
    """Model Predictive Controller Solver
    
    Solves the optimal control problem:
        min J = sum(||x_i - x_ref||_Q^2 + ||u_i||_R^2)
        s.t. x_{i+1} = f(x_i, u_i)
             x, u constraints
    """
    
    def __init__(self, config: MPCConfig):
        """Initialize MPC solver
        
        Args:
            config: MPC configuration
        """
        self.config = config
        self.model = BicycleModel(config)
        self.constraints = ConstraintSet(config)
        
        # Store last solution for warm-starting
        self.last_control_sequence = np.zeros((config.horizon, 2))
        self.last_trajectory = None
        
        # For debugging
        self.last_solve_info = {}
    
    def solve(
        self,
        x0: np.ndarray,
        reference_trajectory: np.ndarray,
        x0_control: Optional[np.ndarray] = None
    ) -> Tuple[np.ndarray, np.ndarray, dict]:
        """Solve MPC optimization problem
        
        Args:
            x0: Initial state [x, y, theta, delta]
            reference_trajectory: Reference trajectory (horizon+1, 4)
            x0_control: Initial control guess (horizon, 2), uses last if None
        
        Returns:
            optimal_controls: Control sequence (horizon, 2)
            predicted_trajectory: Predicted trajectory (horizon+1, 4)
            info: Dictionary with solver info (success, iterations, etc.)
        """
        # Use warm start or provided initial guess
        if x0_control is None:
            u0 = self.last_control_sequence.flatten()
        else:
            u0 = x0_control.flatten()
        
        # Build optimization problem
        bounds = self._build_bounds()
        
        # Objective function
        def objective(u_flat):
            trajectory = self.model.predict_trajectory(x0, u_flat.reshape(-1, 2))
            cost = self._compute_cost(trajectory, u_flat.reshape(-1, 2), reference_trajectory)
            return cost
        
        # Solve
        result = minimize(
            objective,
            u0,
            method='SLSQP',
            bounds=bounds,
            options={'maxiter': 100, 'ftol': 1e-6}
        )
        
        # Extract solution
        optimal_controls = result.x.reshape(-1, 2)
        predicted_trajectory = self.model.predict_trajectory(x0, optimal_controls)
        
        # Store for warm-starting next solve
        self.last_control_sequence = optimal_controls
        self.last_trajectory = predicted_trajectory
        
        # Info dict
        info = {
            'success': result.success,
            'iterations': result.nit,
            'function_value': result.fun,
            'message': result.message
        }
        self.last_solve_info = info
        
        return optimal_controls, predicted_trajectory, info
    
    def get_control(
        self,
        x0: np.ndarray,
        reference_trajectory: np.ndarray
    ) -> np.ndarray:
        """Get first control input of optimal sequence (receding horizon)
        
        Args:
            x0: Current state
            reference_trajectory: Reference trajectory (horizon+1, 4)
        
        Returns:
            control: First control [v, delta_dot]
        """
        controls, _, _ = self.solve(x0, reference_trajectory)
        return controls[0]
    
    def _build_bounds(self) -> list:
        """Build parameter bounds for optimizer
        
        Returns:
            bounds: List of (min, max) tuples for each control variable
        """
        u_lower, u_upper = self.constraints.get_input_bounds()
        
        bounds = []
        for i in range(self.config.horizon):
            bounds.append((u_lower[i, 0], u_upper[i, 0]))  # velocity
            bounds.append((u_lower[i, 1], u_upper[i, 1]))  # steering rate
        
        return bounds
    
    def _compute_cost(
        self,
        trajectory: np.ndarray,
        controls: np.ndarray,
        reference_trajectory: np.ndarray
    ) -> float:
        """Compute total cost
        
        Args:
            trajectory: Predicted trajectory (N+1, 4)
            controls: Control sequence (N, 2)
            reference_trajectory: Reference trajectory (N+1, 4)
        
        Returns:
            cost: Total cost value
        """
        cost = 0.0
        N = len(controls)
        
        # Stage costs
        for i in range(N):
            # State tracking error
            error = trajectory[i] - reference_trajectory[i]
            error[2] = wrap_angle(error[2])  # Wrap angle error
            
            state_cost = error @ self.config.Q @ error
            
            # Control effort
            control_cost = controls[i] @ self.config.R @ controls[i]
            
            cost += state_cost + control_cost
        
        # Terminal cost
        terminal_error = trajectory[-1] - reference_trajectory[-1]
        terminal_error[2] = wrap_angle(terminal_error[2])
        terminal_cost = terminal_error @ self.config.Q_terminal @ terminal_error
        cost += terminal_cost
        
        return cost
    
    def set_weights(self, Q: np.ndarray, R: np.ndarray, Q_terminal: np.ndarray = None):
        """Update cost weights
        
        Useful for tuning controller online
        
        Args:
            Q: State cost matrix
            R: Control cost matrix
            Q_terminal: Terminal cost matrix
        """
        self.config.Q = Q
        self.config.R = R
        if Q_terminal is not None:
            self.config.Q_terminal = Q_terminal
        else:
            self.config.Q_terminal = Q * 2
    
    def reset_warm_start(self):
        """Reset warm-start storage (useful after large state jumps)"""
        self.last_control_sequence = np.zeros((self.config.horizon, 2))
        self.last_trajectory = None
